pkg/component/runtime/service.go

```diff
@@ -0,0 +1,441 @@
+		} else if s.missedCheckins > 0 && s.missedCheckins < maxCheckinMisses {
+			s.compState(client.UnitStateDegraded, cli)
+		} else if s.missedCheckins >= maxCheckinMisses {
+			// something is wrong; the service should be checking in
+			msg := fmt.Sprintf("Failed: %s service missed %d check-ins", cli, maxCheckinMisses)
+			s.forceCompState(client.UnitStateFailed, msg)
+		}
+	}
+	s.lastServiceStatus = status
+}
+
+// Watch returns the channel that sends component state.
+//
+// Channel should send a new state anytime a state for a unit or the whole component changes.
+func (s *ServiceRuntime) Watch() <-chan ComponentState {
+	return s.ch
+}
+
+// Start starts the service.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Start() error {
+	s.actionCh <- serviceStart
+	return nil
+}
+
+// Update updates the currComp runtime with a new-revision for the component definition.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Update(comp component.Component) error {
+	s.compCh <- comp
+	return nil
+}
+
+// Stop stops the service.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Stop() error {
+	s.actionCh <- serviceStop
+	return nil
+}
+
+// Teardown stop and uninstall the service.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Teardown() error {
+	s.actionCh <- serviceTeardown
+	return nil
+}
+
+func (s *ServiceRuntime) forceCompState(state client.UnitState, msg string) {
+	if s.state.forceState(state, msg) {
+		s.sendObserved()
+	}
+}
+
+func (s *ServiceRuntime) sendObserved() {
+	s.ch <- s.state.Copy()
+}
+
+// start starts the service
+func (s *ServiceRuntime) start(ctx context.Context, cli service.Service) error {
+	s.forceCompState(client.UnitStateStarting, fmt.Sprintf("Starting: %s service", cli))
+
+	// Call the check command of the service
+	err := s.check(ctx)
+	if err != nil {
+		// Check failed, call the install command of the service
+		s.log.Errorf("failed check %s service: %v, try install", cli, err)
+		err = s.install(ctx)
+		if err != nil {
+			return fmt.Errorf("failed install %s service: %w", cli, err)
+		}
+	}
+
+	// Check service status if it's already running
+	status, err := cli.Status()
+	if err != nil {
+		return fmt.Errorf("failed checking %s service status: %w", cli, err)
+	}
+
+	// Service is running, set state and return
+	if status == service.StatusRunning {
+		s.forceCompState(client.UnitStateHealthy, fmt.Sprintf("Healthy: %s service is running", cli))
+		return nil
+	}
+
+	// Start the service
+	err = cli.Start()
+	if err != nil {
+		return fmt.Errorf("failed starting %s service: %w", cli, err)
+	}
+
+	// Check service status after attempting it to start
+	status, err = cli.Status()
+	if err != nil {
+		return fmt.Errorf("failed checking %s service status: %w", cli, err)
+	}
+
+	if status == service.StatusRunning {
+		s.forceCompState(client.UnitStateHealthy, fmt.Sprintf("Healthy: %s service is running", cli))
+	}
```

Comment:
What if the status is now not running? Should a different state be set?

Reply:
the status will be rechecked on checking timer and updated accordingly